OSDOCS-2457: document procedures for Global Options to Enable HTTP Strict Transport Security (HSTS) (#35812)
openshift/openshift-docs: modules/nw-enforcing-hsts-per-domain.adoc

+
[source,terminal]
----
$ oc annotate route --all --all-namespaces --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
----
+
* To apply HSTS to all routes in a particular namespace, enter the `oc annotate command`. For example:
+
[source,terminal]
----
$ oc annotate route --all -n my-namespace --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
----

.Verification

You can review the HSTS policy you configured. For example:

* To review the `maxAge` set for required HSTS policies, enter the following command:
+
[source,terminal]
----
$ oc get clusteroperator/ingress -n openshift-ingress-operator -o jsonpath='{range .spec.requiredHSTSPolicies[*]}{.spec.requiredHSTSPolicies.maxAgePolicy.largestMaxAge}{"\n"}{end}'
----
+
* To review the HSTS annotations on all routes, enter the following command:
+
[source,terminal]
----
$ oc get route  --all-namespaces -o go-template='{{range .items}}{{if .metadata.annotations}}{{$a := index .metadata.annotations "haproxy.router.openshift.io/hsts_header"}}{{$n := .metadata.name}}{{with $a}}Name: {{$n}} HSTS: {{$a}}{{"\n"}}{{else}}{{""}}{{end}}{{end}}{{end}}'

> Name: <_routename_> HSTS: max-age=31536000;preload;includeSubDomains

Reviewer: I think you want a separate code block with a `.Example output` block title.

// Module included in the following assemblies:
// * networking/configuring-routing.adoc

[id="nw-enforcing-hsts-per-domain_{context}"]
= Enforcing HTTP strict transport security per-domain

> To enforce HTTP strict transport security (HSTS) per-domain for all TLS routes, add a `requiredHSTSPolicies` record to the Ingress spec to capture the configuration of the HSTS policy.

Reviewer 2: ```suggestion
To enforce HTTP Strict Transport Security (HSTS) per-domain for secure routes, add a `requiredHSTSPolicies` record to the Ingress spec to capture the configuration of the HSTS policy.
```

> If you configure the `route.openshift.io/RequiredRouteAnnotations` plug-in to enforce HSTS, then any newly created route must be configured with a compliant HSTS policy, which is verified against the global setting on the cluster Ingress configuration, called `ingresses.config.openshift.io/cluster`.

Reviewer 2: ```suggestion
If you configure a `requiredHSTSPolicy` to enforce HSTS, then any newly created route must be configured with a compliant HSTS policy annotation.
```

[NOTE]
====
To handle upgraded clusters with non-compliant HSTS routes, you can update the manifests at the source and apply the updates.
====

[IMPORTANT]
====

> HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.

Reviewer 2: Sometimes we say secure, other times we say TLS, probably using secure instead of TLS is better.  Or just pick the one that corresponds with other docs.

====

.Prerequisites

* You are logged in to the cluster with a user with `cluster-admin` privileges.
* You installed the `oc` CLI.

.Procedure

. Edit the Ingress config file:
+
[source,terminal]
----
$ oc edit ingresses.config.openshift.io/cluster
----
+
.Example HSTS policy
[source,yaml]
----
apiVersion: config.openshift.io/v1
kind: Ingress
metadata:
  name: cluster
spec:
  domain: apps.abc.com
  requiredHSTSPolicies: <1>
  - domainPatterns: <2>
    - abc.com
    namespaceSelector: <3>
      matchLabels:
        myPolicy: strict
    maxAge: <4>
      smallestMaxAge: 1
      largestMaxAge: 31536000
    preloadPolicy: RequirePreload <5>

>     includeSubDomainsPolicy: RequireIncludeSubDomains

Reviewer 2: We need a <6> for includeSubDomainsPolicy

Reviewer 2: It can be left blank, or options are: RequireIncludeSubDomains, RequireNoIncludeSubDomains, or NoOpinion

----

> <1> Required. `requiredHSTSPolicies` are validated in order, and the first matching `domainPatterns` applies.

Reviewer 2: ```suggestion
<1> Required. `requiredHSTSPolicies` are validated in order, and the first matching domain in `domainPatterns` applies.  At least one domain must be listed in `domainPatterns`, but any number of domains can be listed.
```

> <2> Required. You must specify at least one `domainPatterns` hostname.

Reviewer 2: ```suggestion
<2> Required. You must specify at least one `domainPatterns` hostname.  Any number of domains can be listed.
```

> <3> Optional. If you include `namespaceSelector`, it must match the labels of the route namespace.

Reviewer 2: ```suggestion
<3> Optional. If you include `namespaceSelector`, it must match the labels of the route's namespace in order to enforce HSTS on routes in that namespace.  
```

> <4> Required. `max-age` measures the length of time, in seconds, that the HSTS policy is in effect. The client updates `max-age` whenever a response with a HSTS header is received from the host. When `max-age` times out, the client discards the policy.

Reviewer 2: ```suggestion
<4> Required. `max-age` measures the length of time, in seconds, that the HSTS policy is in effect. This policy setting allows for a smallest and largest `max-age` to be enforced.  
```

Reviewer 2: The `largestMaxAge` value can be left unspecified, in which case no upper limit is enforced, but it must be between 0 and 2147483647.
The `smallestMaxAge` value can be left unspecified, in which case no lower limit is enforced, but it must be between 0 and 2147483647.  Normally choose 0 if you want to allow HSTS to be disabled for troubleshooting, otherwise choose 1 if you never want HSTS to be disabled.

> <5> Optional. Including `preload` in `haproxy.router.openshift.io/hsts_header` when `max-age` is greater than `0` allows external services to include this site in their HSTS preload lists. For example, sites such as Google can construct a list of sites that have `preload` set. Browsers can then use these lists to determine which sites they can communicate with over HTTPS, before they have interacted with the site. Without `preload` set, browsers must have interacted with the site over HTTPS to get the header.

Reviewer 2: ```suggestion
<5> Optional. Including `preload` in `haproxy.router.openshift.io/hsts_header` allows external services to include this site in their HSTS preload lists. Browsers can then use these lists to determine which sites they can communicate with over HTTPS, before they have interacted with the site. Without `preload` set, browsers may need to interact at least once with the site to get the header.
```

Reviewer 2: Options are: RequirePreload, RequireNoPreload, or NoOpinion

+
. You can apply HSTS to all routes in the cluster or in a particular namespace by entering the `oc annotate command`.
+
* To apply HSTS to all routes in the cluster, enter the `oc annotate command`. For example:
+
[source,terminal]
----
$ oc annotate route --all --all-namespaces --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
----
+
* To apply HSTS to all routes in a particular namespace, enter the `oc annotate command`. For example:
+
[source,terminal]
----
$ oc annotate route --all -n my-namespace --overwrite=true "haproxy.router.openshift.io/hsts_header"="max-age=31536000"
----

.Verification

> You can review the HSTS configurations guaranteed by the admission plugin.

Reviewer 2: ```suggestion
You can review the HSTS policy you configured.
```

> * To review the `maxAge` set for required HSTS policies, enter the following command:

Reviewer 2: ```suggestion
* For example, to review the `maxAge` set for required HSTS policies, enter the following command:
```
